### PR TITLE
Add iis2mdc multi instance support

### PR DESCRIPTION
**Pull request by avisconti**
IIS2MDC sensor

1. Add multi-instance support
2. Remove SPI 4-wires (FIX)

Tested on IKS02A1 shield

EDIT:
I made use of new helpers introduced in #30536 (DEVICE_DT_GET, SPI_CONFIG_DT_INST).
The SPI part was tested in #32874, with a different sensor (iis2iclx) but with same approach

### CODE REVIEW

File: drivers/sensor/iis2mdc/iis2mdc.c
Comment (gmarull) [inline drivers/sensor/iis2mdc/iis2mdc.c:333]: This happens to all Zephyr drivers, but I haven't seen such kind of warning. I think it should be discussed with maintainers, if found useful it could be made available somehow to all drivers.

File: drivers/sensor/iis2mdc/iis2mdc.c
Comment (gmarull) [inline drivers/sensor/iis2mdc/iis2mdc.c:389]: I'd personally choose something like:
```c
		.spi_cs_label = UTIL_AND(                       \
			DT_INST_SPI_DEV_HAS_CS_GPIOS(n),        \
			DT_INST_SPI_DEV_CS_GPIOS_LABEL(n)),     \
		.spi_cs_pin = UTIL_AND(                         \
			DT_INST_SPI_DEV_HAS_CS_GPIOS(n),        \
			DT_INST_SPI_DEV_CS_GPIOS_PIN(n)),       \
		.spi_cs_flags = UTIL_AND(                       \
			DT_INST_SPI_DEV_HAS_CS_GPIOS(n),        \
			DT_INST_SPI_DEV_CS_GPIOS_FLAGS(n)),     \
```

in order to reduce the number of macros. non-blocking, though

Comment (pabigot) [reply drivers/sensor/iis2mdc/iis2mdc.c:389]: I though @mbolivar-nordic had something in works or merged to deal with this?  Yes, it's some commits in #30536.

Maybe we should get that merged first; it's got a lot of stuff not directly related to BME280.

Comment (gmarull) [reply drivers/sensor/iis2mdc/iis2mdc.c:389]: @pabigot you're right. If #30536 is going to be merged soon I'd wait for it and rework this PR using the new helpers.

Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc.c:389]: #30536 looks interesting. I see some macros reusable in this context (SPI_CS_CONTROL_XXX).

Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc.c:389]: We can wait for it to be merged first.

Comment (mbolivar-nordic) [reply drivers/sensor/iis2mdc/iis2mdc.c:389]: It's merged now, feel free to update this PR.

Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc.c:389]: > It's merged now, feel free to update this PR.

Sure, I'll work on it asap.

**Comment by avisconti** [comment]
This PR has been simplified using new DT helpers as reported in #32874.
Just for your info.

**Review by gmarull — APPROVED** [review]

File: drivers/sensor/iis2mdc/iis2mdc_i2c.c
Comment (MaureenHelm) [inline drivers/sensor/iis2mdc/iis2mdc_i2c.c:45]: Replace commas with semicolons

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (MaureenHelm) [inline drivers/sensor/iis2mdc/iis2mdc_spi.c:56]: Did you mean to remove this check?

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (MaureenHelm) [inline drivers/sensor/iis2mdc/iis2mdc_spi.c:89]: Did you mean to remove this check?

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (MaureenHelm) [inline drivers/sensor/iis2mdc/iis2mdc_spi.c:91]: Stray addition? I don't see this getting used anywhere.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/iis2mdc/iis2mdc_i2c.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_i2c.c:45]: Yes, it has been corrected on all drivers. But this commit was pending since a lot of time and I forgot it. Thanks for catching it.

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:56]: Yes, because it should be not there. But, as already discussed with @pabigot in another PR, this will be done in a separate PR for all ST sensor drivers later on. So, I leave it in here.

Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:56]: I mean, I'll leave the check in. Thanks again for catching it.

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:91]: They are indeed used in the stmemsc module. This required by the stmemsc HAL i/f to map the hardware specific i2c/spi methods:

https://github.com/zephyrproject-rtos/hal_st/blob/master/sensor/stmemsc/iis2mdc_STdC/driver/iis2mdc_reg.h

```
typedef struct {
      /** Component mandatory fields **/
     stmdev_write_ptr  write_reg;                                                                                                                             
     stmdev_read_ptr   read_reg;
     /** Customizable optional pointer **/
     void *handle;
  } stmdev_ctx_t;
```

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:89]: see above

File: drivers/sensor/iis2mdc/iis2mdc.c
Comment (mbolivar-nordic) [inline drivers/sensor/iis2mdc/iis2mdc.c:345]: This is still using label properties. You can get the device directly and avoid other boilerplate by using `GPIO_DT_SPEC_GET(DT_DRV_INST(inst), drdy_gpios)`. See the mesh_demo sample for examples.

File: drivers/sensor/iis2mdc/iis2mdc.h
Comment (mbolivar-nordic) [inline drivers/sensor/iis2mdc/iis2mdc.h:45]: Can just make this a `struct gpio_dt_spec` instead when applying the above suggestion.

**Review by mbolivar-nordic — COMMENTED** [review]
Looks good. There's still one stray use of a GPIO device label you may want to fix up.

File: drivers/sensor/iis2mdc/iis2mdc.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc.c:345]: > This is still using label properties. You can get the device directly and avoid other boilerplate by using `GPIO_DT_SPEC_GET(DT_DRV_INST(inst), drdy_gpios)`. See the mesh_demo sample for examples.

@mbolivar-nordic 
Seems ok, but I need to rebase the PR to latest master branch, as here GPIO_DT_SPEC_GET is not yet available.

File: drivers/sensor/iis2mdc/iis2mdc.h
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc.h]: Yes, done both! Thanks

File: drivers/sensor/iis2mdc/iis2mdc_trigger.c
Comment (mbolivar-nordic) [inline drivers/sensor/iis2mdc/iis2mdc_trigger.c:115]: You also need to update this to use device_is_ready() or device_usable_check() instead. The device pointer can be non-NULL but point to a device which is not initialized.

File: drivers/sensor/iis2mdc/iis2mdc.c
Comment (mbolivar-nordic) [inline drivers/sensor/iis2mdc/iis2mdc.c:346]: Could simplify to

```
#define IIS2MDC_CFG_IRQ(inst) \
    IF_ENABLED(CONFIG_IIS2MDC_TRIGGER, (.gpio_drdy = GPIO_DT_SPEC_GET(DT_DRV_INST(inst), drdy_gpios),))
```

Not blocking on this one

Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc.c:346]: :(

File: drivers/sensor/iis2mdc/iis2mdc_trigger.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_trigger.c:115]: I'll use device_is_ready()

File: drivers/sensor/iis2mdc/iis2mdc.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc.c:346]: Sorry, I'm too tired. It's almost 10pm and I want to close this, as tomorrow I need to do other stuff. Sorry to turn this down. But I really appreciate your helps and guidance on this helpers usage!

File: drivers/sensor/iis2mdc/iis2mdc_trigger.c
Comment (mbolivar-nordic) [reply drivers/sensor/iis2mdc/iis2mdc_trigger.c]: > I'll use device_is_ready()

You need to do this for all the devices you depend on. That includes the bus too.

**Review by mbolivar-nordic — COMMENTED** [review]
-1 until the bus is checked for readiness too

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
clicked wrong button and GitHub is making me put something here

File: drivers/sensor/iis2mdc/iis2mdc_trigger.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_trigger.c]: Rad yesterday evening, but it was too late and my mind was totally blank... :D 
So, I think I'll better move this che and the bus check in the main driver init. I'll check all the devices there (I think I have to check gpio device for drdy, and device for bus. I don't think something else is left).

**Comment by avisconti** [comment]
> -1 until the bus is checked for readiness too

I'm afraid I'm not getting what this point is related to. Are you still talking about device_ready() checks?

File: drivers/sensor/iis2mdc/iis2mdc_trigger.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_trigger.c]: Mmmmmh, and SPI cs gpio_dev too. So, maybe better to keep the checks in different routines! Let me know if I'm missing something again... :(

**Comment by avisconti** [comment]
> Looks good. There's still one stray use of a GPIO device label you may want to fix up.

@mbolivar-nordic.
I re-pushed again, because there were a couple of unused drdy interrupt field in data. Now I use all drdy gpio info from config structure.

**Comment by avisconti** [comment]
@mbolivar-nordic @MaureenHelm Did I address all your request?

**Comment by mbolivar-nordic** [comment]
> @mbolivar-nordic @MaureenHelm Did I address all your request?

We're using -ENODEV for missing devices elsewhere and I think that's better than -EINVAL. Decision belongs to @MaureenHelm .

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (MaureenHelm) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:98]: How does the typedef in the header file show that this instance of the struct is getting used somewhere?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> We're using -ENODEV for missing devices elsewhere and I think that's better than -EINVAL. Decision belongs to @MaureenHelm .

Agree with @mbolivar-nordic . Let's try be consistent

**Comment by avisconti** [comment]
> Agree with @mbolivar-nordic . Let's try be consistent

ok, ack

**Comment by avisconti** [comment]
I now returning -ENODEV if device has not been found.
I will do in a separate PR (and for all pertinent sensor drivers ) following things:
    - use stmemsc common routines
    - use a single ctx declaration
    - make use of new DT helpers (here the use of gpio_dt_spec for dry gpio is still not done)

File: drivers/sensor/iis2mdc/iis2mdc.c
Comment (mbolivar-nordic) [reply drivers/sensor/iis2mdc/iis2mdc.c:324]: > I think it should be discussed with maintainers, if found useful it could be made available somehow to all drivers.

I think it's nicer to proceed incrementally or perhaps treat this as a quality of implementation issue. There are lots of inconsistencies in how we emit warnings and log issues in drivers and I'm not sure it is a good idea to hold up this PR on decisions for how to standardize.

**Review by mbolivar-nordic — APPROVED** [review]
my comments are addressed

**Comment by avisconti** [comment]
@MaureenHelm 
Did I address all the points raised by you?

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:98]: > How does the typedef in the header file show that this instance of the struct is getting used somewhere?

Not sure I got the question properly. Every sensor in the STMEMSC HAL module is using a similar ctx structure that the HAL users (the zephyr ST sensor drivers) need to populate with the platform bus specific read/write routines. The module does not know whether the ctx structure is used somewhere or not. Do you see an issue in that?

Comment (pabigot) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:98]: I think @MaureenHelm's point is that `iis2mdc_spi_ctx` is not referenced anywhere.  Unless the pointer to it is being constructed from some magic macro catenation?

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (pabigot) [inline drivers/sensor/iis2mdc/iis2mdc_spi.c:114]: Should this be:

    data->ctx = &iis2mdc_spi_ctx;

?

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:98]: Ah, yes. Now I see. This declaration is not needed anymore. Thanks for catching it.
Sorry for confusion.

Comment (pabigot) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:98]: This and all the `stmdev_ctx_t` instances in the existing drivers should be `static`: there's no need to access them by name from outside the driver implementation.

Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:98]: I mean, THIS declaration:

```
stmdev_ctx_t iis2mdc_spi_ctx = {
	.read_reg = (stmdev_read_ptr) iis2mdc_spi_read,
	.write_reg = (stmdev_write_ptr) iis2mdc_spi_write,
};
```

Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c]: re-pushed. Thanks again!

**Review by MaureenHelm — APPROVED** [review]

File: drivers/sensor/iis2mdc/iis2mdc_spi.c
Comment (avisconti) [reply drivers/sensor/iis2mdc/iis2mdc_spi.c:109]: @pabigot 
No, I just removed the iis2mdc_spi_ctx because the ctx structure is directly inside data. But is like this since long back, and all structures like iis2mdc_spi_ctx and iis2mdc_i2c_ctx are not needed anymore. I'm not sure how it happened that in this PR at a certain point I brought in again this structure, but at least I learned the term "stray addition". Thanks for catching it.
BTW, in ii2smdc I still need to use the stmemsc coomon routines introduced in #33506. Moreover, I'm considering moving the ctx inside stmemsc_cfg_xxx structures, so that I should be able to completely remove the ST_DRVXXX_i2c.c and ST_DRVXXX_spi.c files completely.

**Pull request merged**
